lib/psci/psci_suspend.c

```diff
@@ -85,6 +85,14 @@ static void psci_suspend_to_pwrdown_start(unsigned int end_pwrlvl,
 	 */
 	cm_init_my_context(ep);
 
+	/*
+	 * Plat. management: Allow the platform to perform any early
+	 * actions required to power down the CPU, before we disable
+	 * the data cache.
+	 */
+	if (psci_plat_pm_ops->pwr_domain_suspend_early)
+		psci_plat_pm_ops->pwr_domain_suspend_early(state_info);
+
```

Comment:
Sorry, I didn't properly think through  the full implication of moving this to psci_cpu_suspend_start() earlier. Typically we dont want to tear down platform resources for power down without invoking the SPD hook.  So the correct place for the new hook would be after the `psci_spd_pm->svc_suspend()` SPD hook.

Thinking more about this, my opinion is that this new hook should be an optional hook whereas `pwr_domain_suspend()` would still be the mandatory hook for CPU_SUSPEND.  This is because ,

1. Platforms which require interconnect programming would need to do the same after caches are disabled.
2. The sequence limitation of calling SPD hook prior to any platform programming for suspend restricts the new hook to just power down suspend and therefore the `psci_spd_pm->svc_suspend()` needs to be used for `pure` retention suspend. The new hook is not useful for `pure` retention suspends anyway.
3. Having 2 hooks for doing platform suspend operations would be confusing as a Platform API IMO. Hence maintaining the current hook as mandatory while reducing the new hook as an optimization opportunity is less confusing I think.

If we go with what I mentioned, this new hook can be renamed to `psci_plat_pm_ops->pwr_domain_suspend_pwrdown_early` or something similar.